lib/matplotlib/colorizer.py

```diff
@@ -0,0 +1,486 @@
+
+  :ref:`colormap-manipulation` for examples of how to make
+  colormaps.
+
+  :ref:`colormaps` an in-depth discussion of choosing
+  colormaps.
+
+  :ref:`colormapnorms` for more details about data normalization.
+"""
+
+import numpy as np
+from numpy import ma
+import functools
+from matplotlib import _api, colors, cbook, scale, cm
+
+
+class Colorizer():
+    """
+    Class that holds the data to color pipeline
+    accessible via `Colorizer.__call__(A)` and executed via
+    the `Colorizer.norm` and `Colorizer.cmap` attributes.
+    """
+    def __init__(self, cmap=None, norm=None):
+
+        self._cmap = None
+        self._set_cmap(cmap)
+
+        self._id_norm = None
+        self._norm = None
+        self.norm = norm
+
+        self.callbacks = cbook.CallbackRegistry(signals=["changed"])
+        self.colorbar = None
+
+    def _scale_norm(self, norm, vmin, vmax, A):
+        """
+        Helper for initial scaling.
+
+        Used by public functions that create a ScalarMappable and support
+        parameters *vmin*, *vmax* and *norm*. This makes sure that a *norm*
+        will take precedence over *vmin*, *vmax*.
+
+        Note that this method does not set the norm.
+        """
+        if vmin is not None or vmax is not None:
+            self.set_clim(vmin, vmax)
+            if isinstance(norm, colors.Normalize):
+                raise ValueError(
+                    "Passing a Normalize instance simultaneously with "
+                    "vmin/vmax is not supported.  Please pass vmin/vmax "
+                    "directly to the norm when creating it.")
+
+        # always resolve the autoscaling so we have concrete limits
+        # rather than deferring to draw time.
+        self.autoscale_None(A)
+
+    @property
+    def norm(self):
+        return self._norm
+
+    @norm.setter
+    def norm(self, norm):
+        _api.check_isinstance((colors.Normalize, str, None), norm=norm)
+        if norm is None:
+            norm = colors.Normalize()
+        elif isinstance(norm, str):
+            try:
+                scale_cls = scale._scale_mapping[norm]
+            except KeyError:
+                raise ValueError(
+                    "Invalid norm str name; the following values are "
+                    f"supported: {', '.join(scale._scale_mapping)}"
+                ) from None
+            norm = _auto_norm_from_scale(scale_cls)()
+
+        if norm is self.norm:
+            # We aren't updating anything
+            return
+
+        in_init = self.norm is None
+        # Remove the current callback and connect to the new one
+        if not in_init:
+            self.norm.callbacks.disconnect(self._id_norm)
+        self._norm = norm
+        self._id_norm = self.norm.callbacks.connect('changed',
+                                                    self.changed)
+        if not in_init:
+            self.changed()
+
+    def __call__(self, x, alpha=None, bytes=False, norm=True):
```

Comment:
Given that we are almost always going to use this in library code (and only type it out in a terminal rarely) I am weakly against using `__call__`.  That we can do this is Python is sometimes very nice from an API point of view, but it is mostly in the case where you want to hide the fact that you up-converted from a plain function (with usually no state) to an object that holds a bunch of state (or is acting as a "mutable closure").

Even though we use this pattern with norm and colormap, I think we should break it here and give this a name!

Reply:
:smile: Almost the same points I made [above](https://github.com/matplotlib/matplotlib/pull/28658#discussion_r1710353662). Let's give it a name.

Reply:
I changed it back to `to_rgba`, since that is already in use in the library, I think that makes the most sense